lib/costellation.dart

```diff
@@ -0,0 +1,642 @@
+  }
+
+  bool hasItem(Directory directory) {
+    Pointer<Char> item = directory.name.toNativeUtf8().cast<Char>();
+
+    bool has_item = bindings.directory_has_item(_pointer, item);
+
+    if (has_item == 0) {
+      return false;
+    }
+    calloc.free(item);
+    return true;
+  }
+
+  void addFile(File file) {
+    G_FFIResult_Null result =
+        bindings.directory_add_file(_pointer, file._pointer);
+
+    if (result.error != nullptr) {
+      throw WarpException(result.error);
+    }
+    bindings.free(result.data);
+  }
+
+  void addItem(Item item) {
+    G_FFIResult_Null result =
+        bindings.directory_add_item(_pointer, item.pointer());
+
+    if (result.error != nullptr) {
+      throw WarpException(result.error);
+    }
+    bindings.free(result.data);
+  }
+
+  Item directoryToItem() {
+    Pointer<G_Item> result = bindings.directory_into_item(_pointer);
+
+    if (result == nullptr) {
+      throw Exception("File not found");
+    }
+    Item item = Item(result);
+    bindings.item_free(result);
+    return item;
+  }
+
+  Directory setDescription(String descritpion) {
+    Pointer<G_Item> item = bindings.directory_into_item(_pointer);
+
+    if (item == nullptr) {
+      throw Exception("File not found");
+    }
+
+    bindings.item_set_description(
+        item, descritpion.toNativeUtf8().cast<Char>());
+
+    _pointer = bindings.item_into_directory(item).data;
+
+    return Directory(_pointer);
+  }
+
+  Directory rename(String name) {
+    Pointer<G_Item> item = bindings.directory_into_item(_pointer);
+
+    if (item == nullptr) {
+      throw Exception("File not found");
+    }
+
+    bindings.item_rename(item, name.toNativeUtf8().cast<Char>());
+
+    _pointer = bindings.item_into_directory(item).data;
+
+    return Directory(_pointer);
+  }
+
+  Item getItem(String itemS) {
+    G_FFIResult_Item result = bindings.directory_get_item(
+        _pointer, itemS.toNativeUtf8().cast<Char>());
+
+    if (result.error != nullptr) {
+      throw WarpException(result.error);
+    }
+    Item item = Item(result.data);
+
+    bindings.item_free(result.data);
+    return item;
+  }
+
+  int getItemIndex(String item) {
+    G_FFIResult_usize result = bindings.directory_get_item_index(
+        _pointer, item.toNativeUtf8().cast<Char>());
+
+    if (result.error != nullptr) {
+      throw WarpException(result.error);
+    }
+    int index = result.data.value;
+    calloc.free(result.data);
+    return index;
+  }
+
+  List<Item> listItems() {
+    Pointer<G_FFIVec_Item> pointerListItem =
+        bindings.directory_get_items(_pointer);
+    if (pointerListItem == nullptr) {
+      throw Exception("Items or directory not found");
+    }
+    List<Item> list = [];
+    int length = pointerListItem.ref.len;
+
+    for (int i = 0; i < length; i++) {
+      Pointer<G_Item> pointer = pointerListItem.ref.ptr.elementAt(i).value;
+      Item key = Item(pointer);
+      list.add(key);
+      bindings.item_free(pointer);
```

Comment:
Freeing should be omitted here since `Item` class will be holding onto the pointer, otherwise it would cause a segment fault when attempting to access the pointer